pkg/monitor/azure/nsg/nsg.go

```diff
@@ -0,0 +1,155 @@
+
+	metricInvalidDenyRule         = "preconfigurednsg.invalid.denyrule"
+	metricSubnetAcessResponseCode = "preconfiguredNSG.subnetaccess.responsecode"
+	metricSubnetAccessForbidden   = "preconfigurednsg.subnetaccess.forbidden"
+
+	expandNSG = "NetworkSecurityGroup"
+)
+
+// NSGMonitor is responsible for performing NSG rule validations when preconfiguredNSG is enabled
+type NSGMonitor struct {
+	log     *logrus.Entry
+	emitter metrics.Emitter
+	oc      *api.OpenShiftCluster
+
+	subnetCli network.SubnetsClient
+}
+
+func NewNSGMonitor(log *logrus.Entry, oc *api.OpenShiftCluster, subscriptionID string, environment *azureclient.AROEnvironment, cred refreshable.Authorizer, emitter metrics.Emitter) *NSGMonitor {
+	return &NSGMonitor{
+		log:     log,
+		emitter: emitter,
+		oc:      oc,
+
+		subnetCli: network.NewSubnetsClient(environment, subscriptionID, cred),
+	}
+}
+
+type subnetNSGConfig struct {
+	// subnet CIDR range
+	prefix netip.Prefix
+	// The rules from the subnet NSG
+	nsg *sdknetwork.SecurityGroup
+}
+
+func (n *NSGMonitor) toSubnetConfig(ctx context.Context, subnetID string) (subnetNSGConfig, error) {
+	r, err := arm.ParseResourceID(subnetID)
+	if err != nil {
+		return subnetNSGConfig{}, err
+	}
+
+	dims := map[string]string{
+		dimSubnet:         r.Name,
+		dimVnet:           r.Parent.Name,
+		dimResourceGroup:  r.ResourceGroupName,
+		dimSubscriptionID: r.SubscriptionID,
+	}
+
+	subnet, err := n.subnetCli.Get(ctx, r.ResourceGroupName, r.Parent.Name, r.Name, expandNSG)
+	// response code is always returned as part of autorest
+	n.emitter.EmitGauge(metricSubnetAcessResponseCode, int64(subnet.StatusCode), dims)
+	if err != nil {
+		if subnet.StatusCode == http.StatusForbidden {
+			n.emitter.EmitGauge(metricSubnetAccessForbidden, int64(1), dims)
+		}
+		n.log.Errorf("Error while getting subnet %s. %s", subnetID, err)
+		return subnetNSGConfig{}, err
+	}
+
+	cidr, err := toPrefix(*subnet.AddressPrefix)
+	if err != nil {
+		return subnetNSGConfig{}, err
+	}
+	return subnetNSGConfig{cidr, subnet.NetworkSecurityGroup}, nil
+}
+
+func (n *NSGMonitor) Monitor(ctx context.Context) error {
+	masterSubnet, err := n.toSubnetConfig(ctx, n.oc.Properties.MasterProfile.SubnetID)
+	if err != nil {
+		// FP has no access to the subnet
+		return err
+	}
+
+	workerSubnets := make([]subnetNSGConfig, 0, len(n.oc.Properties.WorkerProfiles))
+	workerPrefixes := make([]netip.Prefix, 0, len(n.oc.Properties.WorkerProfiles))
+	for _, wp := range n.oc.Properties.WorkerProfiles {
+		s, err := n.toSubnetConfig(ctx, wp.SubnetID)
+		if err != nil {
+			// FP has no access to the subnet
+			return err
+		}
+		workerSubnets = append(workerSubnets, s)
+		workerPrefixes = append(workerPrefixes, s.prefix)
+	}
+
+	// to make sure each NSG is processed only once
+	nsgSet := map[string]*sdknetwork.SecurityGroup{
+		*masterSubnet.nsg.ID: masterSubnet.nsg,
+	}
+	for _, w := range workerSubnets {
+		nsgSet[*w.nsg.ID] = w.nsg
+	}
+
+	for nsgID, nsg := range nsgSet {
```

Comment:
Just curious - do we have different rules for worker NSG vs. master NSG?

Reply:
With BYO-NSG, I think it's possible.

Reply:
Basically each subnet can be attached with a different NSG altogether.